internal/cmd/cmd.go

```diff
@@ -197,49 +210,73 @@ func (c *Cmd) Start() error {
+		// c.Stdin.CloseWrite() (if it implements it) or c.Stdin.Close().
+		// However, we are only passed on the Reader end of Stdin, and closing the
+		// upstream c.Stdin breaks with the functionality that os.exec.Cmd implements.
 		go func() {
-			_, err := relayIO(stdin, c.Stdin, c.Log, "stdin")
+			_, err := relayIO(stdin, c.Stdin, c.Log, "stdin", isIOChannelClosedErr)
 			// Report the stdin copy error. If the process has exited, then the
 			// caller may never see it, but if the error was due to a failure in
 			// stdin read, then it is likely the process is still running.
 			if err != nil {
-				c.stdinErr.Store(err)
+				select {
+				case <-c.allDoneCh:
+					// Wait has returned, err will be ignored.
+					// relayIO will log the error, so no need for logging here
+				default:
+					c.stdinErr.Store(err)
+				}
 			}
-			// Notify the process that there is no more input.
-			if err := p.CloseStdin(context.TODO()); err != nil && c.Log != nil {
+			// Notify the process that there is no more input, in the case that
+			// c.Stdin closed while the process is running
+			if err := p.CloseStdin(ctx); err != nil && !isIOChannelClosedErr(err) && c.Log != nil {
 				c.Log.WithError(err).Warn("failed to close Cmd stdin")
 			}
 		}()
 	}
 
 	if c.Stdout != nil {
 		c.iogrp.Go(func() error {
-			_, err := relayIO(c.Stdout, stdout, c.Log, "stdout")
-			if err := p.CloseStdout(context.TODO()); err != nil {
-				c.Log.WithError(err).Warn("failed to close Cmd stdout")
+			_, err := relayIO(c.Stdout, stdout, c.Log, "stdout", nil /*skipErr*/)
+			// Notify the process that upstream IO closed its std out, if the process
+			// is still running
+			if cerr := p.CloseStdout(ctx); cerr != nil && !isIOChannelClosedErr(cerr) && c.Log != nil {
+				c.Log.WithError(cerr).Warn("failed to close Cmd stdout")
 			}
 			return err
 		})
 	}
 
 	if c.Stderr != nil {
 		c.iogrp.Go(func() error {
-			_, err := relayIO(c.Stderr, stderr, c.Log, "stderr")
-			if err := p.CloseStderr(context.TODO()); err != nil {
-				c.Log.WithError(err).Warn("failed to close Cmd stderr")
+			_, err := relayIO(c.Stderr, stderr, c.Log, "stderr", nil /*skipErr*/)
+			// Notify the process that upstream IO closed its std err, if the process
+			// is still running
+			if cerr := p.CloseStderr(ctx); cerr != nil && !isIOChannelClosedErr(cerr) && c.Log != nil {
+				c.Log.WithError(cerr).Warn("failed to close Cmd stderr")
 			}
 			return err
 		})
 	}
 
+	// if ctx is `Background()`, then don't bother launching this, since ctx will
+	// never be cancelled
 	if c.Context != nil {
 		go func() {
 			select {
 			case <-c.Context.Done():
-				_, _ = c.Process.Kill(context.TODO())
+				// Process.Kill (via Process.Signal) will not send an RPC if the context passed
+				// in is cancelled (ie, bridge.AsyncRPC returns early)
+				// todo: create context with span info, but does not inherit cancellation
+				_, _ = c.Process.Kill(context.Background())
```

Comment:
Should this be `ctx`?

Reply:
not in this case, since if the context is cancelled, the process may still be running, but the bridge wont send the kill/signal RPC using a done context. So we have to pass in a context that we know is not cancelled to make sure the signal is sent

Reply:
We should add a comment about that